Projects: Publishing A/B test log to firehose

Erin007 opened the pull request:
Follow up to #24257 and #24291. 

This PR logs to Redshift when a the user (un)publishes via the buttons in the button variant of the project publishing A/B test and when the user (un)publishes via the quick action in the dropdown menu in the chevron variant. Poorva requested that user id also be included in the logs; part of accomplishing that is passing a `userId` prop into the relevant child components. 

Next, we'd like to set up tracking clicks on the chevron icon itself which will be slightly more complicated as the `QuickActionsCell` is used in multiple tables throughout the site.  

Example data being logged: 
<img width="736" alt="screen shot 2018-08-22 at 10 48 49 am" src="https://user-images.githubusercontent.com/12300669/44480820-15718100-a5f9-11e8-99e7-25d087a9191b.png">

Erin007 commented on apps/src/templates/projects/PersonalProjectsPublishedCell.jsx:
```diff
@@ -16,13 +17,32 @@ class PersonalProjectsPublishedCell extends Component {
+        study: 'project-publish',
+        study_group: 'publish-button',
+        event: 'publish',
+        user_id: this.props.userId,
```
I haven't done this kind of logging yet. 1.) can I just add in extra fields like this? The examples I've found (mostly #22486) don't. 2.) Is there a way to check locally that this logging is working or do I just ask Poorva to check Redshift once the change hits staging?

balderdash replied:
1.) So this should actually work, but it was probably not meant to work this way.

You can't normally just add new fields like this, you would first need to add columns to the table that this logs to. The available fields are here:
https://github.com/code-dot-org/code-dot-org/blob/99db4d453e4007150d82d960968f835d9913f5ad/apps/src/lib/util/firehose.js#L10-L20

However `user_id` actually exists already, but it's not meant to be specified in the data argument to the `putRecord` call. It's a field that is automatically added to the firehose record if you include `includeUserId: true` in the options argument to `putRecord`: https://github.com/code-dot-org/code-dot-org/blob/99db4d453e4007150d82d960968f835d9913f5ad/apps/src/lib/util/firehose.js#L183

Unfortunately, you then run into the problem that the `pageConstants` reducer isn't used on non-puzzle pages, so the firehose code can't actually look up the user_id when you're in the project gallery:
https://github.com/code-dot-org/code-dot-org/blob/99db4d453e4007150d82d960968f835d9913f5ad/apps/src/lib/util/firehose.js#L159-L162

The way that `addCommonValues` is written though, your code should actually just work. It won't touch your manually specified `user_id`, and firehose/redshift won't know the difference between this and a normal automatically populated `user_id`.

I think you have two options:
1. Add the pageConstants reducer to the redux store in the project gallery, and populate the user_id so that the `includeUserId` option works as intended.
2. Update the usage comment for `putRecord` to officially allow a `user_id` property (but also mention the `includeUserId` option in the comment), and update the code above so that it doesn't override a manually specified `user_id` with the one from redux. You probably want to do this.

balderdash replied:
2.) I think that you can pass the `alwaysPut` option to `putRecord` to force it to log to firehose in development mode, but you'd still have to ask Poorva to check redshift to see if the record was added (or set up your own access to redshift). ~~If my memory serves, you also have to wait a day or so for the record to actually show up.~~ Turns out it shows up in Redshift within minutes.

I usually don't bother with that. In development mode `putRecord` logs to the JS console instead of to firehose, I just make sure it does that.

balderdash commented on apps/src/templates/projects/PersonalProjectsPublishedCell.jsx:
```diff
@@ -16,13 +17,32 @@ class PersonalProjectsPublishedCell extends Component {
+        study_group: 'publish-button',
+        event: 'unpublish',
+        user_id: this.props.userId,
+        channel_id: this.props.projectId,
```
For new fields other than `user_id`, I've just used the `data_json` field, so something like:
```
data_json: JSON.stringify({ channel_id: this.props.projectId }),
```

Erin007 commented:
@balderdash I added in explanatory comments regarding user_id, put channel_id in the data blob and updated the description to include confirmation that the expected log entry is being outputted. Anything else you recommend?
